src/ports/postgres/modules/dbscan/dbscan.py_in

```diff
@@ -209,8 +521,187 @@ def dbscan_predict(schema_madlib, dbscan_table, source_table, id_column,
             """.format(**locals())
         result = plpy.execute(sql)
 
+def rtree_transition(state, id_in, expr_points, eps, min_samples, metric, n_rows, leaf_id, **kwargs):
+
+    SD = kwargs['SD']
+    if not state:
+        data = {}
+        SD['counter{0}'.format(leaf_id)] = 0
+    else:
+        data = SD['data{0}'.format(leaf_id)]
+
+    data[id_in] = expr_points
+    SD['counter{0}'.format(leaf_id)] = SD['counter{0}'.format(leaf_id)]+1
+    SD['data{0}'.format(leaf_id)] = data
+    ret = [[-1,-1],[-1,-1]]
+
+    my_n_rows = n_rows[leaf_id]
+
+    if SD['counter{0}'.format(leaf_id)] == my_n_rows:
+
+        core_counts = {}
+        core_lists = {}
+        p = index.Property()
+        p.dimension = len(expr_points)
+        idx = index.Index(properties=p)
+        ret = []
+
+        if metric == 'dist_norm1':
+            fn_dist = distance.cityblock
+        elif metric == 'dist_norm2':
+            fn_dist = distance.euclidean
+        else:
+            fn_dist = distance.sqeuclidean
+
+        for key1, value1 in data.items():
+            idx.add(key1,value1+value1,key1)
```

Comment:
That will require transferring the rtree structure from one step to the next. We can use pickle to get a bytea but I think it would take some additional time to pickle/unpickle the whole tree for each row.

Reply:
We already save the points in SD between each row.  This would just be saving it as a kdtree data structure instead of a list data structure.  (Or if we also need the list, we could save both.). No pickling necessary since it never gets transferred between segments or written out to disk.

However--after reading this "streaming optimization" for rtree:
https://rtree.readthedocs.io/en/latest/performance.html#use-stream-loading

I'm wondering if bulk loading is already optimized inside rtree--if that's the case, then it's probably better to just leave it as is.